[Lens as Code] Refactor `colorByValue` schema and transforms

nickofthyme opened the pull request:
## Summary

Simplifies `colorByValue` schema into an easy explicit range configuration.

```ts
// before
const steps = [
  { type: 'from', from: 0, color: 'red' }, // from is only allowed on first step
  { type: 'exact', value: 50, color: 'green' }, // exact is confusing
  { type: 'to', to: 75, color: 'blue' }, // to is only allowed on last step
];

// after
const steps = [
  { lt: 0, color: 'red' }, // gte is optional
  { gte: 0, lt: 50, color: 'green' },
  { gte: 50, lte: 75, color: 'blue' }, // lte/lt is optional
];
```

Additionally, the `ColorStep.stop` value was already being set to `null` in some cases, even though the type is strictly a `number`, same for `rangeMin` and `rangeMax`. These are both updated to permit `null` values to indicate and unbounded step.

Confirmed dynamic by-value color configurations work correctly for both Gauge and Metric charts.

### Changes

- `colorByValue` schema
  - remove `to`, `from` and `exact` step types
  - merge `colorByValueSchema` into single schema
  - remove top-level `min` and `max` properties
  - add single `step` type with optional `gte`, `lt` and `lte`
  - limit min size of steps to `1`
  - add validation for step configurations
  - update color schema tests
 - Color transforms
  - update transform logic to use new `gte`/`lt`/`lte` steps
  - convert `reverse` steps into normalized steps
  - add validation test cases for mixed bounds usage

<details><summary>See config comparisons</summary>
<p>

| Colors in UI | Old config | New Config |
|--------------|------------|------------|
| <img width="355" height="203" alt="image" src="https://github.com/user-attachments/assets/a3847292-8100-42b7-8238-645ad7bbe3fe" /> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "continuity": "none",<br>  "steps": [<br>    {<br>      "type": "from",<br>      "color": "green",<br>      "from": 1570<br>    },<br>    {<br>      "type": "exact",<br>      "color": "yellow",<br>      "value": 3140<br>    },<br>    {<br>      "type": "to",<br>      "color": "red",<br>      "to": 4710<br>    }<br>  ]<br>}</pre> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "steps": [<br>    {<br>      "gte": 0,<br>      "lt": 1570,<br>      "color": "green"<br>    },<br>    {<br>      "gte": 1570,<br>      "lt": 3140,<br>      "color": "yellow"<br>    },<br>    {<br>      "gte": 3140,<br>      "lte": 4710,<br>      "color": "red"<br>    }<br>  ]<br>}</pre> |
| <img width="349" height="199" alt="image" src="https://github.com/user-attachments/assets/f67d9107-1adc-416d-9bc0-8657bc335b3a" /> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "continuity": "below",<br>  "steps": [<br>    {<br>      "type": "from",<br>      "color": "green",<br>      "from": 1570<br>    },<br>    {<br>      "type": "exact",<br>      "color": "yellow",<br>      "value": 3140<br>    },<br>    {<br>      "type": "to",<br>      "color": "red",<br>      "to": 4710<br>    }<br>  ]<br>}</pre> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "steps": [<br>    {<br>      "lt": 1570,<br>      "color": "green"<br>    },<br>    {<br>      "gte": 1570,<br>      "lt": 3140,<br>      "color": "yellow"<br>    },<br>    {<br>      "gte": 3140,<br>      "lte": 4710,<br>      "color": "red"<br>    }<br>  ]<br>}</pre> |
| <img width="349" height="196" alt="image" src="https://github.com/user-attachments/assets/a17d598f-807d-4e38-a6e3-eba0c4667b7c" /> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "continuity": "all",<br>  "steps": [<br>    {<br>      "type": "from",<br>      "color": "green",<br>      "from": 1570<br>    },<br>    {<br>      "type": "exact",<br>      "color": "yellow",<br>      "value": 3140<br>    },<br>    {<br>      "type": "to",<br>      "color": "red",<br>      "to": 4710<br>    }<br>  ]<br>}</pre> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "steps": [<br>    {<br>      "lt": 1570,<br>      "color": "green"<br>    },<br>    {<br>      "gte": 1570,<br>      "lt": 3140,<br>      "color": "yellow"<br>    },<br>    {<br>      "gte": 3140,<br>      "color": "red"<br>    }<br>  ]<br>}</pre> |
| <img width="350" height="204" alt="image" src="https://github.com/user-attachments/assets/fe44bbe2-05a6-42fd-b8c5-87610ae1680a" /> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "continuity": "above",<br>  "steps": [<br>    {<br>      "type": "from",<br>      "color": "green",<br>      "from": 1570<br>    },<br>    {<br>      "type": "exact",<br>      "color": "yellow",<br>      "value": 3140<br>    },<br>    {<br>      "type": "to",<br>      "color": "red",<br>      "to": 4710<br>    }<br>  ]<br>}</pre> | <pre>{<br>  "type": "dynamic",<br>  "range": "absolute",<br>  "steps": [<br>    {<br>      "gte": 0,<br>      "lt": 1570,<br>      "color": "green"<br>    },<br>    {<br>      "gte": 1570,<br>      "lt": 3140,<br>      "color": "yellow"<br>    },<br>    {<br>      "gte": 3140,<br>      "color": "red"<br>    }<br>  ]<br>}</pre> |

</p>
</details> 

Fix #251621

## Details

We found and issue with the legacy color `stops` as detailed in #251135. Because of this we need a way to return the original (i.e. incorrect) color stops to the visualization in certain cases.

This issue has no impact on `custom` palettes.

For all non-custom palettes we use the `'legacy-dynamic'` palette type. With this we check if a shift if needed and encode this into the config in order to output the correct original palette. This new type looks like this...

```tsx
export const legacyColorByValueSchema = colorByValueBaseSchema.extends({
    type: schema.literal('legacy-dynamic'),
    palette: schema.string(),
    shift: schema.boolean(),
});
```

Ideally the root issue (#251135) is fixed before GA and we can remove this support. Worst case, we leave it in as deprecated and eventually fix the root issue.

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

nickofthyme committed:
refactor: colorByValue types

- remove `to`, `from` and `exact` step types
- merge `colorByValueSchema` into single schema
- remove top-level `min` and `max` properties
- add single `step` type with optional `to` and `from`
- limit min size of steps to `1`
- add validation for step configurations
- update color schema tests

nickofthyme committed:
update color transform logic with `to`, `from` steps

nickofthyme committed:
chore: cleanup color types and fix type errors

nickofthyme committed:
test: add dynamic color metric example mock

nickofthyme committed:
fix: last step issues and update coloring tests

nickofthyme committed:
chore: add reverse logic converting state to api

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368795)
* Commit: 5659ed9a43ee47a9cdf23b7ea4b401182849f2a2

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/368795#019ae795-1d26-4993-a07f-4a909f40a3f4)

cc @nickofthyme

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368795","number":368795,"commit":"5659ed9a43ee47a9cdf23b7ea4b401182849f2a2"}],"number":368795}
buildkite-pr-comment-->

markov00 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/color.ts:
```diff
@@ -9,124 +9,92 @@
+      )
+    ),
+    /**
+     * The value up to which this color applies (inclusive).
```
Looking at the current UI all the steps except the last follow this logic `gte - lt`, so the `to` is exclusive and the form is inclusive. The last one is `gte - lte`. We should probably be explicit with that somehow

In the screenshot below the ranges are:
- `gte: 0 lt: 20`
- `gte: 20 lt: 80`
- `gte:80 lte: 100`


<img width="462" height="248" alt="Image" src="https://github.com/user-attachments/assets/4a8806e9-4842-4285-9105-1b9239e02466" />

markov00 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/color.ts:
```diff
@@ -9,124 +9,92 @@
+     * The value from which this color applies (inclusive).
+     */
+    from: schema.maybe(
+      schema.nullable(
```
I'm wondering why both undefined or nullable? can we just set to one or the other type of non-defined?

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
refactor: color schema to use `gte`, `lte` and `lt`

nickofthyme committed:
chore: update coloring transforms with new step schema

nickofthyme committed:
test: update chart type transform mocks

nickofthyme committed:
chore: fix color union schema

nickofthyme committed:
fix: more test cases with new color config

nickofthyme committed:
chore: revert type changes to fix later

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
test: cleanup, fix and add types to schema tests

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/gauge.test.ts:
```diff
@@ -235,14 +240,16 @@ describe('Gauge Schema', () => {
+        type: 'invalid',
       },
-    };
+    } satisfies GaugeInput;
```
I think it's also good to type the invalid cases then explicitly use `@ts-expect-error` where the type is deliberately invalid.

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/371504)
* Commit: 895409753ed9465f3f0ac7b08bfbee72eac8682d

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/da2623f5c805b21d4e72ed6f917669edcc70740e) | [after](https://github.com/elastic/kibana/commit/895409753ed9465f3f0ac7b08bfbee72eac8682d) | diff |
| --- | --- | --- | --- |
| `dashboard` | 979 | 980 | +1 |
| `discover` | 1874 | 1875 | +1 |
| `infra` | 1753 | 1754 | +1 |
| `lens` | 1651 | 1652 | +1 |
| `observability` | 1647 | 1648 | +1 |
| `observabilityAIAssistantApp` | 778 | 779 | +1 |
| `onechat` | 1542 | 1543 | +1 |
| `unifiedDocViewer` | 769 | 770 | +1 |
| total |  |  | +8 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/da2623f5c805b21d4e72ed6f917669edcc70740e) | [after](https://github.com/elastic/kibana/commit/895409753ed9465f3f0ac7b08bfbee72eac8682d) | diff |
| --- | --- | --- | --- |
| `dashboard` | 702.8KB | 702.7KB | -102.0B |
| `discover` | 1.3MB | 1.3MB | -98.0B |
| `infra` | 1.1MB | 1.1MB | -102.0B |
| `lens` | 1.6MB | 1.6MB | -99.0B |
| `observability` | 1.7MB | 1.7MB | -98.0B |
| `observabilityAIAssistantApp` | 258.3KB | 258.2KB | -104.0B |
| `onechat` | 479.0KB | 478.9KB | -102.0B |
| `unifiedDocViewer` | 364.3KB | 364.2KB | -106.0B |
| total |  |  | -811.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/da2623f5c805b21d4e72ed6f917669edcc70740e) | [after](https://github.com/elastic/kibana/commit/895409753ed9465f3f0ac7b08bfbee72eac8682d) | diff |
| --- | --- | --- | --- |
| `lens` | 64.4KB | 64.4KB | -2.0B |
| `unifiedDocViewer` | 13.4KB | 13.4KB | -1.0B |
| total |  |  | -3.0B |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/da2623f5c805b21d4e72ed6f917669edcc70740e) | [after](https://github.com/elastic/kibana/commit/895409753ed9465f3f0ac7b08bfbee72eac8682d) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-common` | 0 | 1 | +1 |
| `@kbn/lens-embeddable-utils` | 21 | 40 | +19 |
| `exploratoryView` | 22 | 24 | +2 |
| `expressionHeatmap` | 0 | 1 | +1 |
| `lens` | 87 | 88 | +1 |
| `securitySolution` | 434 | 435 | +1 |
| `visTypeTimeseries` | 31 | 32 | +1 |
| `visualizations` | 21 | 22 | +1 |
| total |  |  | +27 |


</details>


### History
* :broken_heart: [Build #371379](https://buildkite.com/elastic/kibana-pull-request/builds/371379) failed c379a56d9a42a31fb15811d75c9ece7d77145143

cc @nickofthyme

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371504","number":371504,"commit":"895409753ed9465f3f0ac7b08bfbee72eac8682d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371379","number":371379,"commit":"c379a56d9a42a31fb15811d75c9ece7d77145143"}],"number":371504}
buildkite-pr-comment-->

dej611 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/color.test.ts:
```diff
@@ -7,115 +7,224 @@
+          steps: [
+            { gte: 0, lt: 50, color: 'red' },
+            { gte: 50, lt: 75, color: 'green' },
+            { gte: 75, lte: 100, color: 'blue' },
```
If `lt` of the current step and `gte` of the ext step must be exactly the same, why can't just we drop the `lt` one in this case?

Take this for example, it looks clearer than the one above who leaks some implementation details:

```ts
[
  { gte: 0, color: 'red' },
  { gte: 50, color: 'green' },
  { gte: 75, lte: 100, color: 'blue' }
]
```

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
fix: new schema color configurations

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/377624)
* Commit: ec46373b39e103914264b87da3d693cd7cb33fd7

### Failed CI Steps
* [Scout: [ platform / streams_app ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/377624#019b80d9-937d-471a-9520-4fdaddddc532)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/c260e0ba11e3b0399568875f96337a5a29babca7) | [after](https://github.com/elastic/kibana/commit/ec46373b39e103914264b87da3d693cd7cb33fd7) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1394 | 1395 | +1 |
| `dashboard` | 996 | 997 | +1 |
| `discover` | 1950 | 1951 | +1 |
| `infra` | 1772 | 1773 | +1 |
| `lens` | 1661 | 1662 | +1 |
| `observability` | 1688 | 1689 | +1 |
| `observabilityAIAssistantApp` | 800 | 801 | +1 |
| `unifiedDocViewer` | 778 | 779 | +1 |
| total |  |  | +8 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/c260e0ba11e3b0399568875f96337a5a29babca7) | [after](https://github.com/elastic/kibana/commit/ec46373b39e103914264b87da3d693cd7cb33fd7) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 528.7KB | 528.6KB | -102.0B |
| `dashboard` | 721.4KB | 721.3KB | -100.0B |
| `discover` | 1.3MB | 1.3MB | -105.0B |
| `infra` | 1.1MB | 1.1MB | -102.0B |
| `lens` | 1.9MB | 1.9MB | +534.0B |
| `observability` | 1.7MB | 1.7MB | -102.0B |
| `observabilityAIAssistantApp` | 267.9KB | 267.8KB | -103.0B |
| `unifiedDocViewer` | 371.0KB | 370.9KB | -104.0B |
| total |  |  | -184.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/c260e0ba11e3b0399568875f96337a5a29babca7) | [after](https://github.com/elastic/kibana/commit/ec46373b39e103914264b87da3d693cd7cb33fd7) | diff |
| --- | --- | --- | --- |
| `dashboard` | 17.5KB | 17.5KB | +1.0B |
| `observabilityAIAssistantApp` | 17.4KB | 17.4KB | +1.0B |
| total |  |  | +2.0B |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/c260e0ba11e3b0399568875f96337a5a29babca7) | [after](https://github.com/elastic/kibana/commit/ec46373b39e103914264b87da3d693cd7cb33fd7) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-common` | 0 | 1 | +1 |
| `@kbn/lens-embeddable-utils` | 36 | 57 | +21 |
| `exploratoryView` | 22 | 24 | +2 |
| `expressionHeatmap` | 0 | 1 | +1 |
| `lens` | 87 | 88 | +1 |
| `securitySolution` | 438 | 439 | +1 |
| `visTypeTimeseries` | 31 | 32 | +1 |
| `visualizations` | 21 | 22 | +1 |
| total |  |  | +29 |


</details>


### History
* :broken_heart: [Build #377603](https://buildkite.com/elastic/kibana-pull-request/builds/377603) failed 7ffc474b4064cf55bcee2da6cdad6ffa7e4ec89f
* :green_heart: [Build #371504](https://buildkite.com/elastic/kibana-pull-request/builds/371504) succeeded 895409753ed9465f3f0ac7b08bfbee72eac8682d
* :broken_heart: [Build #371379](https://buildkite.com/elastic/kibana-pull-request/builds/371379) failed c379a56d9a42a31fb15811d75c9ece7d77145143

cc @nickofthyme

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377624","number":377624,"commit":"ec46373b39e103914264b87da3d693cd7cb33fd7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377603","number":377603,"commit":"7ffc474b4064cf55bcee2da6cdad6ffa7e4ec89f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371504","number":371504,"commit":"895409753ed9465f3f0ac7b08bfbee72eac8682d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371379","number":371379,"commit":"c379a56d9a42a31fb15811d75c9ece7d77145143"}],"number":377624}
buildkite-pr-comment-->

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/378372)
* Commit: 0cead11c3f7a1626a549064e61203c3bb5bde453

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/59c6276e3a68313d54d38ba51ef80ab6a03de9db) | [after](https://github.com/elastic/kibana/commit/0cead11c3f7a1626a549064e61203c3bb5bde453) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1396 | 1397 | +1 |
| `dashboard` | 1020 | 1021 | +1 |
| `discover` | 1976 | 1977 | +1 |
| `infra` | 1892 | 1893 | +1 |
| `lens` | 1677 | 1678 | +1 |
| `observability` | 1804 | 1805 | +1 |
| `observabilityAIAssistantApp` | 800 | 801 | +1 |
| `unifiedDocViewer` | 778 | 779 | +1 |
| total |  |  | +8 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/59c6276e3a68313d54d38ba51ef80ab6a03de9db) | [after](https://github.com/elastic/kibana/commit/0cead11c3f7a1626a549064e61203c3bb5bde453) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 528.9KB | 528.8KB | -102.0B |
| `dashboard` | 785.0KB | 784.9KB | -100.0B |
| `discover` | 1.4MB | 1.4MB | -105.0B |
| `infra` | 1.2MB | 1.2MB | -102.0B |
| `lens` | 1.9MB | 1.9MB | +534.0B |
| `observability` | 1.8MB | 1.8MB | -102.0B |
| `observabilityAIAssistantApp` | 267.9KB | 267.8KB | -103.0B |
| `unifiedDocViewer` | 371.1KB | 371.0KB | -104.0B |
| total |  |  | -184.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/59c6276e3a68313d54d38ba51ef80ab6a03de9db) | [after](https://github.com/elastic/kibana/commit/0cead11c3f7a1626a549064e61203c3bb5bde453) | diff |
| --- | --- | --- | --- |
| `dashboard` | 17.3KB | 17.3KB | +1.0B |
| `observabilityAIAssistantApp` | 17.4KB | 17.4KB | +1.0B |
| total |  |  | +2.0B |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/59c6276e3a68313d54d38ba51ef80ab6a03de9db) | [after](https://github.com/elastic/kibana/commit/0cead11c3f7a1626a549064e61203c3bb5bde453) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-common` | 0 | 1 | +1 |
| `@kbn/lens-embeddable-utils` | 36 | 57 | +21 |
| `exploratoryView` | 22 | 24 | +2 |
| `expressionHeatmap` | 0 | 1 | +1 |
| `lens` | 87 | 88 | +1 |
| `securitySolution` | 438 | 439 | +1 |
| `visTypeTimeseries` | 31 | 32 | +1 |
| `visualizations` | 21 | 22 | +1 |
| total |  |  | +29 |


</details>


### History
* :yellow_heart: [Build #377624](https://buildkite.com/elastic/kibana-pull-request/builds/377624) was flaky ec46373b39e103914264b87da3d693cd7cb33fd7
* :broken_heart: [Build #377603](https://buildkite.com/elastic/kibana-pull-request/builds/377603) failed 7ffc474b4064cf55bcee2da6cdad6ffa7e4ec89f
* :green_heart: [Build #371504](https://buildkite.com/elastic/kibana-pull-request/builds/371504) succeeded 895409753ed9465f3f0ac7b08bfbee72eac8682d
* :broken_heart: [Build #371379](https://buildkite.com/elastic/kibana-pull-request/builds/371379) failed c379a56d9a42a31fb15811d75c9ece7d77145143

cc @nickofthyme

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/378372","number":378372,"commit":"0cead11c3f7a1626a549064e61203c3bb5bde453"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377624","number":377624,"commit":"ec46373b39e103914264b87da3d693cd7cb33fd7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377603","number":377603,"commit":"7ffc474b4064cf55bcee2da6cdad6ffa7e4ec89f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371504","number":371504,"commit":"895409753ed9465f3f0ac7b08bfbee72eac8682d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371379","number":371379,"commit":"c379a56d9a42a31fb15811d75c9ece7d77145143"}],"number":378372}
buildkite-pr-comment-->

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/378905)
* Commit: 76837b789ba686833c1caff9953399acdd2a66f5

### Failed CI Steps
* [FTR Configs #34](https://buildkite.com/elastic/kibana-pull-request/builds/378905#019b9b09-176f-4d0c-a63a-8bcbf8b67586)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/378905#019b9b09-176f-4d0c-a63a-8bcbf8b67586) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/378905/jobs/019b9b09-176f-4d0c-a63a-8bcbf8b67586/artifacts/019b9b34-9cfd-48e0-ae8f-64dcd162977a) FTR Configs #<span></span>34 / Endpoint Exceptions with feature flag enabled @<span></span>ess @<span></span>serverless @<span></span>skipInServerlessMKI Endpoint List API (deprecated): RBAC and Validation "before all" hook in "@<span></span>ess @<span></span>serverless @<span></span>skipInServerlessMKI Endpoint List API (deprecated): RBAC and Validation"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/321364b77412099274610dba64af2716794d69ae) | [after](https://github.com/elastic/kibana/commit/76837b789ba686833c1caff9953399acdd2a66f5) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1400 | 1401 | +1 |
| `dashboard` | 1020 | 1021 | +1 |
| `discover` | 1976 | 1977 | +1 |
| `infra` | 1892 | 1893 | +1 |
| `lens` | 1679 | 1680 | +1 |
| `observability` | 1804 | 1805 | +1 |
| `observabilityAIAssistantApp` | 800 | 801 | +1 |
| `unifiedDocViewer` | 778 | 779 | +1 |
| total |  |  | +8 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/321364b77412099274610dba64af2716794d69ae) | [after](https://github.com/elastic/kibana/commit/76837b789ba686833c1caff9953399acdd2a66f5) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 530.4KB | 530.3KB | -102.0B |
| `dashboard` | 784.7KB | 784.6KB | -100.0B |
| `discover` | 1.4MB | 1.4MB | -105.0B |
| `infra` | 1.2MB | 1.2MB | -102.0B |
| `lens` | 1.9MB | 1.9MB | +534.0B |
| `observability` | 1.8MB | 1.8MB | -102.0B |
| `observabilityAIAssistantApp` | 267.9KB | 267.8KB | -103.0B |
| `unifiedDocViewer` | 371.1KB | 371.0KB | -104.0B |
| total |  |  | -184.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/321364b77412099274610dba64af2716794d69ae) | [after](https://github.com/elastic/kibana/commit/76837b789ba686833c1caff9953399acdd2a66f5) | diff |
| --- | --- | --- | --- |
| `dashboard` | 17.3KB | 17.3KB | +1.0B |
| `observabilityAIAssistantApp` | 17.4KB | 17.4KB | +1.0B |
| total |  |  | +2.0B |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/321364b77412099274610dba64af2716794d69ae) | [after](https://github.com/elastic/kibana/commit/76837b789ba686833c1caff9953399acdd2a66f5) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-common` | 0 | 1 | +1 |
| `@kbn/lens-embeddable-utils` | 36 | 57 | +21 |
| `exploratoryView` | 22 | 24 | +2 |
| `expressionHeatmap` | 0 | 1 | +1 |
| `lens` | 87 | 88 | +1 |
| `securitySolution` | 438 | 439 | +1 |
| `visTypeTimeseries` | 31 | 32 | +1 |
| `visualizations` | 21 | 22 | +1 |
| total |  |  | +29 |


</details>


### History
* :green_heart: [Build #378372](https://buildkite.com/elastic/kibana-pull-request/builds/378372) succeeded 0cead11c3f7a1626a549064e61203c3bb5bde453
* :yellow_heart: [Build #377624](https://buildkite.com/elastic/kibana-pull-request/builds/377624) was flaky ec46373b39e103914264b87da3d693cd7cb33fd7
* :broken_heart: [Build #377603](https://buildkite.com/elastic/kibana-pull-request/builds/377603) failed 7ffc474b4064cf55bcee2da6cdad6ffa7e4ec89f
* :green_heart: [Build #371504](https://buildkite.com/elastic/kibana-pull-request/builds/371504) succeeded 895409753ed9465f3f0ac7b08bfbee72eac8682d
* :broken_heart: [Build #371379](https://buildkite.com/elastic/kibana-pull-request/builds/371379) failed c379a56d9a42a31fb15811d75c9ece7d77145143

cc @nickofthyme

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/378905","number":378905,"commit":"76837b789ba686833c1caff9953399acdd2a66f5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/378372","number":378372,"commit":"0cead11c3f7a1626a549064e61203c3bb5bde453"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377624","number":377624,"commit":"ec46373b39e103914264b87da3d693cd7cb33fd7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377603","number":377603,"commit":"7ffc474b4064cf55bcee2da6cdad6ffa7e4ec89f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371504","number":371504,"commit":"895409753ed9465f3f0ac7b08bfbee72eac8682d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371379","number":371379,"commit":"c379a56d9a42a31fb15811d75c9ece7d77145143"}],"number":378905}
buildkite-pr-comment-->

andrimal reviewed (commented):
While testing locally, I can still see some issues with the coloring transformations. 

- Gauge: 

https://github.com/user-attachments/assets/9065780b-dfbe-4685-954f-c053a4210a1f

- Legacy Metric:

https://github.com/user-attachments/assets/c81a5ef3-f79a-4811-b95c-322ba82d0acb

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
chore: quick fixes for pr comments

nickofthyme committed:
fix: initialize gauge ranges with open range

nickofthyme commented on src/platform/packages/shared/kbn-coloring/src/palettes/types.ts:
I fixed the Gauge issues you noticed, the main issue was that we generally initialize color by value stops with open ranges such that there is no min or max. But this is all messed up as we set the `rangeMin` and `rangeMax` to the data domain even when it's set to a percentage range. Also if you go into the default palette it shows the min and max are open but if you save the vis it will lock-in the bad range values.

I don't know what's happened here over the years but I think setting the `rangeMin` and `rangeMax` to `Infinity` should solve the main issues. See fix in 618582ac7440a04548cd64c2a3904ea727e4240e.

nickofthyme committed:
chore: set lens  gauge ranges to `undefined` over `Infinity`

nickofthyme committed:
chore: fix legacy metric default palette configuration to account for correct data bounds

nickofthyme commented on src/platform/packages/shared/kbn-coloring/src/palettes/types.ts:
Next for the issues with the Legacy Metric, same issue as the gauge with the default values of `rangeMin` and `rangeMax` along with other problems.

First off, the default palette sets all the color `stops` to a value of `0` in addition to the `rangeMin` and `rangeMax`, which is not great!

```json
{
  "type": "palette",
  "name": "status",
  "params": {
    "name": "status",
    "reverse": false,
    "rangeType": "number",
    "rangeMin": 0,
    "rangeMax": 0,
    "progression": "fixed",
    "stops": [
      {
        "color": "#24c292",
        "stop": 0
      },
      {
        "color": "#fcd883",
        "stop": 0
      },
      {
        "color": "#f6726a",
        "stop": 0
      }
    ],
    "steps": 3,
    "continuity": "all",
    "maxSteps": 5
  }
}
```

Secondly, the legacy color logic is "shifting" the palette to shift and add a stop at the end for some reason and the palette params are different where the `stops` are the actual numerical value of the stop. But the shift is the main issue in terms of these changes and the errors with transforming.

I think I fixed this in 4aec71e95de3d31ec091dc9ee0a8ec71ba55bbca, but I need to test against previous default palettes, where the user selected color by value but never changes anything, this is the primary issue.

nickofthyme commented on src/platform/packages/shared/kbn-coloring/src/palettes/types.ts:
This issue mainly comes from the `continuity` being set to `above` by default on the gauge. 

But apart from that, looking into the palette logic on `main`, just switching from one palette to another is very bizarre. Specifically in the `changeColorPalette` reducer, the `steps` go in correct, say like this...

```json
[
    {
        "color": "#61a2ff",
        "stop": 25
    },
    {
        "color": "#c8deff",
        "stop": 50
    },
    {
        "color": "#ffccc6",
        "stop": 75
    },
    {
        "color": "#f6726a",
        "stop": 100
    }
]
```

But come out "shifted" for some reason...

```json
[
    {
        "color": "#fcd8d6",
        "stop": 0
    },
    {
        "color": "#feb7b0",
        "stop": 25
    },
    {
        "color": "#fc968d",
        "stop": 50
    },
    {
        "color": "#f6726a",
        "stop": 75
    }
]
```

Even though it's defining the same value ranges.

I think I will put up a PR to fix all of this logic outside of this PR. That way it can be tested in isolation from these changes and test the changes on `main` with and without the Lens transforms.

nickofthyme committed:
chore: update continuity

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
chore: fix more type/schema errors

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
fix bad merge changes to metric.color

nickofthyme committed:
refactor: non-custom palettes should return original stops

- stops are incorrect in a few places
- the ui expects these steps to be incorrect to work as expected
- always handle custom stops correctly
- certain cases should be bwc to return the incorrect old stops
- adds the `palette` property to track the previous palette name to correctly handle bwc.
- shifts bad stops before and after transforms as needed to return previous stops

nickofthyme committed:
chore: add final logic tweaks to correctly capture original palette and needed shifts

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
test: fix errors improve type enforcement

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/metric.ts:
```diff
@@ -221,7 +222,9 @@ const metricStateSecondaryMetricOptionsSchema = {
-  color: schema.maybe(schema.oneOf([colorByValueAbsolute, staticColorSchema])),
+  color: schema.maybe(
+    schema.oneOf([colorByValueAbsoluteSchema, legacyColorByValueAbsoluteSchema, staticColorSchema])
+  ),
```
Not sure about here actually. For the secondary metric I believe we only use the palette name.

andrimal replied:
Exactly 💯 Allowing only static config for the secondary metric color was my thought 
```typescript
color: schema.maybe(staticColorSchema)
``` 

Only one note: I believe we should actually keep the `legacyColorByValueAbsoluteSchema` and use it for the Legacy metic. Currently for the Legacy metric color schema we have: 

```typescript
color: schema.maybe(schema.oneOf([colorByValueAbsoluteSchema, legacyColorByValueSchema]))
``` 

But it actually only allows numeric ranges.

nickofthyme replied:
Good catch, I missed the legacy was only absolute. Fixed in https://github.com/elastic/kibana/pull/245159/commits/38d92b589df0548ed7a5f7b8905fadb932393f1e

markov00 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/metric.test.ts:
```diff
@@ -167,11 +168,9 @@ describe('Metric Schema', () => {
-                  { type: 'from', from: 0, color: '#blue' },
-                  { type: 'to', to: 100, color: '#red' },
+                  { lt: 0, color: 'blue' },
+                  { gte: 100, color: 'red' },
```
In the previous version of this test it was:
```
min: 0,
max: 100,
steps: [
  { type: 'from', from: 0, color: '#blue' },
  { type: 'to', to: 100, color: '#red' },
]
```
That to me doesn't make much sense, I can't understand what those ranges are.
Am I interpreting correctly that the new ranges are not excluding everything from 0 to <100? 
is that wanted? isn't better to pick up a better set of ranges?

markov00 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/legacy_metric.test.ts:
```diff
@@ -153,19 +159,21 @@ describe('Legacy Metric Schema', () => {
           operation: 'sum',
           field: 'sales',
+          // @ts-expect-error
           apply_color_to: 'invalid',
```
This still need to be fixed

markov00 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/datatable.test.ts:
```diff
@@ -687,8 +687,7 @@ describe('Datatable Schema', () => {
                 {
-                  type: 'from',
-                  from: 0,
+                  lt: 0,
```
This is wrong, it should be `gte:0` instead if you want to color black everything from 0 to infinity (as in the main branch)
This error seems to be copied/pased in many tests where you user `lt:0, gte:100` but this config just create colors ourside the range 0,100

markov00 commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/color.ts:
```diff
@@ -9,127 +9,188 @@
+            description: 'The lower bound of range from which this color applies (inclusive).',
+          },
+        })
+      )
```
Why don't we have the `gt` exclusive?

markov00 reviewed (changes requested):
Overall changes looks good to me, but testing it locally with the API enabled I've found an issue: if you create a metric without a dymanic color, then save a dashboard, then update and add a dynamic color, I can't save the dashboard anymore. It looks like what I'm now trying to save is:
```
"color":{"type":"legacy-dynamic","range":"absolute","palette":"status","shift":false,"steps":[{"lt":null,"color":"#24c292"},{"lt":25.33,"color":"#fcd883"},{"gte":25.33,"color":"#f6726a"}]},
```
that contains `lt:null` and cause the `colorByValueStepSchema` schema validation to fail.
We should probably move the validation one level up and allow this case if we consider it a valid one.

nickofthyme committed:
fix bad color ranges

nickofthyme committed:
fix `apply_color_to` in test and other bad color ranges

nickofthyme committed:
cleanup `metric.color` options

nickofthyme committed:
fix issue with `Infinity` as `min` and `max` ranges

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/metric.ts:
Oh right! Thank you I missed that! :)

nickofthyme committed:
fix color schema on legacy metric

nickofthyme committed:
Merge branch 'main' into fix-color-transforms

nickofthyme committed:
fix legacy metric type guard

andrimal reviewed (approved):
Tested and found no more regressions! 🚀 

Gauge and legacy metric continuity issues will be handled in a follow up.

nickofthyme committed:
Merge branch 'main' into fix-color-transforms
